File: lib/src/main/java/io/github/kory33/guardedqueries/core/subqueryentailments/SubqueryEntailmentInstance.java
package io.github.kory33.guardedqueries.core.subqueryentailments;

import com.google.common.collect.ImmutableBiMap;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSet;
import io.github.kory33.guardedqueries.core.formalinstance.FormalInstance;
import uk.ac.ox.cs.pdq.fol.Constant;
import uk.ac.ox.cs.pdq.fol.Variable;

public record SubqueryEntailmentInstance(
        ImmutableMap<Variable, Constant> ruleConstantWitnessGuess,
        ImmutableSet<Variable> coexistentialVariables,
        FormalInstance<LocalInstanceTerm> localInstance,
        ImmutableMap<Variable, LocalInstanceTerm.LocalName> localWitnessGuess,
        ImmutableBiMap<Constant, LocalInstanceTerm.LocalName /* not in the range of localWitnessGuess */> queryConstantEmbedding
) {
}

File: lib/src/main/scala/io/github/kory33/guardedqueries/core/subqueryentailments/SubqueryEntailmentInstance.scala
package io.github.kory33.guardedqueries.core.subqueryentailments

import io.github.kory33.guardedqueries.core.utils.datastructures.BijectiveMap
import uk.ac.ox.cs.pdq.fol.Constant
import uk.ac.ox.cs.pdq.fol.Variable

case class SubqueryEntailmentInstance(
  ruleConstantWitnessGuess: Map[Variable, Constant],
  coexistentialVariables: Set[Variable],
  localInstance: LocalInstance,
  localWitnessGuess: Map[Variable, LocalInstanceTerm.LocalName],
  queryConstantEmbedding: BijectiveMap[Constant, LocalInstanceTerm.LocalName]
) {
  def ruleConstantWitnessGuessAsMapToInstanceTerms
    : Map[Variable, LocalInstanceTerm.RuleConstant] =
    ruleConstantWitnessGuess
      .view.mapValues(LocalInstanceTerm.RuleConstant.apply)
      .toMap

  def withLocalInstance(newLocalInstance: LocalInstance): SubqueryEntailmentInstance =
    this.copy(localInstance =
      newLocalInstance
    )
}
